Create examples of working with Prefect REST APIs

discdiver opened the pull request:
<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.

Happy engineering!
-->

<!-- Include an overview here -->

### Example
<!-- 
Share an example of the change in action.

A code blurb is best. Changes to features should include an example that is executable by a new user.
If changing documentation, a link to a preview of the page is great.
 -->

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  <!-- If you do not have permission to add a label, a maintainer will add one for you -->

WillRaphaelson reviewed (approved):
This is a smart addition Jeff, I could have used this when I first started digging into the routes more.

discdiver commented:
>  I could have used this when I first started digging into the routes more.

Thank you, Will. Hopefully it helps some folks get going with the REST API faster.  

I started going down the road with httpx before I found Mason's Discourse post and it wasn't the most clear how to authenticate with Cloud before I saw Zanie's issue.

billpalombi commented:
Love this

serinamarie commented:
Thank you @discdiver! I've been wanting this. Might also be worth adding an example of using `requests` library:
https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/artifacts/#artifacts-api

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
 - Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interacti ...
 - The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).
 
+## Interacting with the REST API
+
+You have many options to interact with Prefect REST API
+- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
+- use your favorite Python HTTP library such as [requests](https://requests.readthedocs.io/en/latest/) or [httpx](https://www.python-httpx.org/)
+- use an HTTP library in your language of choice
+- use [CURL](https://curl.se/) from the command line 
```
This isn't showing up as bullet points in the preview. Not sure if broken preview:
https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/#interacting-with-the-rest-api

madkinsz commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+WORKSPACE="my_cloud_workspace_can_be_found_in_my_prefect_cloud_url"
+PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT/workspaces/$WORKSPACE"
+PREFECT_API_KEY="my_api_key_goes_here"
+DEPLOYMENT="my-deployment"
```
This might need to be a UUID?

discdiver replied:
It needed a few fixes - changes made.

madkinsz commented on docs/api-ref/rest-api/index.md:
````diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+Here's an example of using CURL to create a flow run with Prefect Cloud:
+
+```bash
````
Should probably hint to use `prefect config view` after logging in to find your API URL

madkinsz commented on docs/api-ref/rest-api/index.md:
````diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+Here's an example of using CURL to create a flow run with Prefect Cloud:
+
+```bash
+ACCOUNT="my_cloud_account_can_be_found_in_my_prefect_cloud_url"
+WORKSPACE="my_cloud_workspace_can_be_found_in_my_prefect_cloud_url"
````
Might want to demo that these are UUIDs then include how to find it outside of the code blurb or with a comment

madkinsz commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
 
+## Interacting with the REST API
+
+You have many options to interact with Prefect REST API
+- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
```
```suggestion
You have many options to interact with Prefect REST API:

- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
```

Need a newline for bullets to render

madkinsz commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+You have many options to interact with Prefect REST API
+- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
+- use your favorite Python HTTP library such as [requests](https://requests.readthedocs.io/en/latest/) or [httpx](https://www.python-httpx.org/)
```
We should capitalize these to match our style guide

serinamarie replied:
I read this as capitalize the first word of each bullet point, not capitalize `requests` 🤷‍♀️

discdiver commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+You have many options to interact with Prefect REST API
+- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
+- use your favorite Python HTTP library such as [requests](https://requests.readthedocs.io/en/latest/) or [httpx](https://www.python-httpx.org/)
```
```suggestion
- use your favorite Python HTTP library such as [Requests](https://requests.readthedocs.io/en/latest/) or [HTTPX](https://www.python-httpx.org/)
```

discdiver commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +16,60 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
 
+## Interacting with the REST API
+
+You have many options to interact with Prefect REST API
```
```suggestion
You have many options to interact with the Prefect REST API
```

discdiver commented:
@serinamarie and @madkinsz thank you for the feedback! Added a Requests example adapted from @serinamarie's example and fixed made several other changes. Verified all examples worked.

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -4,6 +4,8 @@ tags:
     - REST API
     - Prefect Cloud
     - Prefect Server
+    - curl
```
maybe `requests`?

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+You have many options to interact with the Prefect REST API:
+
+- create an instance of [`PrefectClient`](/api-ref/prefect/client/orchestration/#prefect.client.orchestration.PrefectClient) 
+- use your favorite Python HTTP library such as [Requests](https://requests.readthedocs.io/en/latest/) or [HTTPX](https://www.python-httpx.org/)
+- use an HTTP library in your language of choice
+- use [curl]](https://curl.se/) from the command line 
```
These should be capitalized like Zanie mentioned

serinamarie replied:
```suggestion
- use [curl](https://curl.se/) from the command line 
```

billpalombi commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+- use [curl]](https://curl.se/) from the command line 
+
+### PrefectClient with Prefect server
+Here's an example that uses `PrefectClient` with a locally hosted Prefect server:
```
```suggestion
This example uses `PrefectClient` with a locally hosted Prefect server:
```

billpalombi commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+### Requests with Prefect
+
+Here's an example that uses the Requests library with Prefect Cloud to return the five newest artifacts.
```
```suggestion
This example uses the Requests library with Prefect Cloud to return the five newest artifacts.
```

billpalombi commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+
+### curl with Prefect Cloud
+
+Here's an example that uses curl with Prefect Cloud to create a flow run:
```
```suggestion
This example uses curl with Prefect Cloud to create a flow run:
```

billpalombi reviewed (commented):
Minor nits

serinamarie commented on docs/api-ref/rest-api/index.md:
````diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+```bash
+ACCOUNT_ID="abc-my-cloud-account-id-goes-here"
+WORKSPACE_ID="123-my-workspace-id-goes-here"
+PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT_ID/workspaces/$WORKSPACE_ID""
````
```suggestion
PREFECT_API_URL="https://api.prefect.cloud/api/accounts/$ACCOUNT_ID/workspaces/$WORKSPACE_ID"
```

serinamarie commented on docs/api-ref/rest-api/index.md:
````diff
@@ -16,6 +18,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+Note that in this example `--data-raw "{}"` is required and is where you can specify other aspects of the flow run such as the state. Windows users substitute `^` for `\` for line multi-line commands.
+
+
+## Finding your Prefect Cloud details
+
+When working with the Prefect Cloud REST API you will need your Account ID and often need the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:
+
+<div class="terminal">
+```bash
+PREFECT_API_URL='https://api.prefect.cloud/api/accounts/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here'
+PREFECT_API_KEY='123abc_my_api_key_is_here'
+```
+</div>
+
+Alternatively, view your Account ID and Workspace ID in your browser URL. For example: `https://app.prefect.cloud/account/abc-my-account-id-is-here/workspaces/123-my-workspace-id-is-here`. 
````
It would be nice to know this ahead of time

discdiver replied:
Think I should just move that whole section on finding your cloud details above the examples of interacting with the API, @serinamarie ?

serinamarie replied:
I would consider something like:

The [Prefect REST API](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.

Prefect Cloud and a locally hosted Prefect server each provide a REST API.
- Prefect Cloud:
    - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.
    - Finding your Prefect Cloud details
- Locally hosted open-source Prefect server:
    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the /docs endpoint of the [PREFECT_API_URL](https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with prefect server start to access the interactive documentation.
    - The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api-reference/).

discdiver replied:
done

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -16,6 +19,111 @@ Prefect Cloud and a locally hosted Prefect server each provide a REST API.
+- Use an HTTP library in your language of choice
+- Use [curl](https://curl.se/) from the command line 
+
+### PrefectClient with Prefect server
```
```suggestion
### `PrefectClient` with Prefect server
```

serinamarie replied:
Not sure of the docs style guide on headings but in the preview it looks strange to me, specifically "curl with Prefect" without the code dashes on `curl`

discdiver replied:
curl's website rolls similarly, so I think it's ok as is. :)

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -11,13 +11,16 @@ tags:
 # REST API
 
-The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.
+The [Prefect REST API](https://deploy-preview-9661--prefect-docs-preview.netlify.app/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.
```
```suggestion
The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from clients to the Prefect server so that orchestration can be performed. This API is consumed by clients such as the Prefect Python SDK or the server dashboard.
```

serinamarie replied:
why did they all get changed to deploy preview?

discdiver replied:
I believe I copy-pasted your suggested change. Will fix those. Good catch!

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -3,18 +3,129 @@ description: Prefect REST API for interacting with the orchestration engine and
-- Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interacti ...
-- The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).
+- Prefect Cloud:
+    - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.
```
```suggestion
    - Interactive Prefect Cloud REST API documentation is available at [https://app.prefect.cloud/api/docs](https://app.prefect.cloud/api/docs).
```

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -11,13 +11,16 @@ tags:
+    - Interactive Prefect Cloud REST API documentation is available at https://app.prefect.cloud/api/docs.
+    - [Finding your Prefect Cloud details](#finding-your-prefect-cloud-details)
+- Locally hosted open-source Prefect server:
+    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the /docs endpoint of the [PREFECT_API_URL](https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with prefect server start to access the interac ...
```
`prefect server start`

serinamarie replied:
I think it would be easier to grab the actual markdown instead of trying to copy and paste in the UI because the links have changed and the code snippets too

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -3,7 +3,10 @@ description: Prefect REST API for interacting with the orchestration engine and
+    - Prefect server
+    - curl
+    - PrefectClient
+    - Requests
```
maybe add `API reference`?

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -12,9 +15,117 @@ The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from c
 
 Prefect Cloud and a locally hosted Prefect server each provide a REST API.
```
I'm wondering if this sentence is still necessary now

serinamarie replied:
👍

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -3,7 +3,10 @@ description: Prefect REST API for interacting with the orchestration engine and
+    - Requests
 ---
 
 # REST API
```
> The [Prefect REST API](/api-ref/rest-api/)

@discdiver Not your fault but I'm noticing this is a self-link.

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -12,9 +15,117 @@ The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from c
+    - [Interactive Prefect Cloud REST API documentation](https://app.prefect.cloud/api/docs).
+    - [Finding your Prefect Cloud details](#finding-your-prefect-cloud-details)
+- Locally hosted open-source Prefect server:
+    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the `/docs` endpoint of the [PREFECT_API_URL](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
```
```suggestion
    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at `http://localhost:4200/docs` or the `/docs` endpoint of the [PREFECT_API_URL](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
```

it might make sense to at least add dashes so it's not just a bare link

serinamarie replied:
similar to here:
https://deploy-preview-9661--prefect-docs-preview.netlify.app/concepts/infrastructure/?h=localhost#dockercontainer

serinamarie replied:
Also maybe this documentation could be the second bullet point

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -12,9 +15,117 @@ The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from c
+    - [Finding your Prefect Cloud details](#finding-your-prefect-cloud-details)
+- Locally hosted open-source Prefect server:
+    - Interactive REST API documentation for a locally hosted open-source Prefect server is available at http://localhost:4200/docs or the `/docs` endpoint of the [PREFECT_API_URL](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interactive documentation.
+    - [Prefect REST API documentation](/api-ref/rest-api-reference/).
```
```suggestion
    - [Prefect REST API documentation](/api-ref/rest-api-reference/)
```

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -12,9 +15,117 @@ The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from c
-- Interactive REST API documentation for a locally hosted open-source Prefect server is available at <a href="http://localhost:4200/docs" target="_blank">http://localhost:4200/docs</a> or the `/docs` endpoint of the [`PREFECT_API_URL`](/concepts/settings/#prefect_api_url) you have configured to access the server. You must have the server running with `prefect server start` to access the interacti ...
-- The REST API documentation for a locally hosted open-source Prefect server is also available in the [Prefect REST API Reference](/api-ref/rest-api-reference/).
+- Prefect Cloud:
+    - [Interactive Prefect Cloud REST API documentation](https://app.prefect.cloud/api/docs).
```
```suggestion
    - [Interactive Prefect Cloud REST API documentation](https://app.prefect.cloud/api/docs)
```

serinamarie commented on docs/api-ref/rest-api/index.md:
```diff
@@ -12,9 +15,117 @@ The [Prefect REST API](/api-ref/rest-api/) is used for communicating data from c
+
+## Finding your Prefect Cloud details
+
+When working with the Prefect Cloud REST API you will need your Account ID and often need the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:
```
```suggestion
When working with the Prefect Cloud REST API you will need your Account ID and often the Workspace ID for the [workspace](/cloud/workspaces/) you want to interact with. You can find both IDs for a [Prefect profile](/concepts/settings/) in the CLI with `prefect profile inspect my_profile`. This command will also display your [Prefect API key](/cloud/users/api-keys/), as shown below:
```

serinamarie reviewed (approved):
Thanks so much for this @discdiver! It's going to be incredibly helpful 🔥
